### ISSUE TITLE
Datastore delete immediately after modify doesn't work

### ISSUE THREAD

**Issue by bgomberg** [aws-amplify/amplify-ios#1206]
**Describe the bug**
Using datastore, if I perform a delete immediately after calling save() to modify an existing record, the result of the delete is `.success`, but the record still exists.

**To Reproduce**
1. Create a new record (in this case a `Post`) via `Amplify.DataStore.save()`
2. Wait for the new record to sync with the backend
3. Modify the new record (saving it via `Amplify.DataStore.save()`)
4. Immediately after the completion handler for the `save()` is called, delete the new record
5. Wait for things to sync with the backend
6. Observe that the new record was not deleted (but the modification was persisted)

**Expected behavior**
I would expect the row to have been deleted.

**Environment(please complete the following information):**
 - Amplify Framework Version: 1.9.0
 - Amplify Categories: API + DataStore
 - Dependency Manager: Cocoapods
 - Swift Version: 5.4
 - CLI Version: 4.50.0
 - Xcode Version: 12.5

**Device Information (please complete the following information):**
 - Device: iPhone SE
 - iOS Version: 14.5

**Additional context**
Here's an example unit test which reproduces this issue (based on the types from the getting started guide):
```swift
    func testSaveModifyDelete() throws {
        print("Clearing local DB")
        let clearExpectation = expectation(description: "Clear")
        Amplify.DataStore.clear() { result in
            switch result {
            case .success:
                clearExpectation.fulfill()
            case .failure(let error):
                XCTFail("Error saving post: \(error)")
            }
        }
        wait(for: [clearExpectation], timeout: 10)

        var post = Post(title: UUID().uuidString, status: .published, rating: 5)
        print("Creating post (id=\(post.id), title=\(post.title))")
        let saveExpectation = expectation(description: "Save")
        Amplify.DataStore.save(post) { result in
            switch result {
            case .success:
                saveExpectation.fulfill()
            case .failure(let error):
                XCTFail("Error saving post: \(error)")
            }
        }
        wait(for: [saveExpectation], timeout: 10)

        print("Syncing to cloud")
        sleep(10)

        // Reading posts
        print("Querying posts")
        let readExpectation = expectation(description: "Read")
        Amplify.DataStore.query(Post.self) { result in
            switch result {
            case .success(let posts):
                XCTAssertEqual(posts.count, 1)
                XCTAssertEqual(posts.first!.id, post.id)
                readExpectation.fulfill()
            case .failure(let error):
                XCTFail("Error reading post \(error)")
            }
        }
        wait(for: [readExpectation], timeout: 10)

        print("Updating post")
        post.rating = 3
        let updateExpectation = expectation(description: "Delete")
        Amplify.DataStore.save(post) { result in
            switch result {
            case .success:
                updateExpectation.fulfill()
            case .failure(let error):
                XCTFail("Error saving post: \(error)")
            }
        }
        wait(for: [updateExpectation], timeout: 10)

        print("Deleting post")
        let deleteExpectation = expectation(description: "Delete")
        Amplify.DataStore.delete(post) { result in
            switch result {
            case .success:
                deleteExpectation.fulfill()
            case .failure(let error):
                XCTFail("Error saving post: \(error)")
            }
        }
        wait(for: [deleteExpectation], timeout: 10)

        print("Syncing to cloud")
        sleep(10)

        print("Reading posts")
        let readExpectation2 = expectation(description: "Read2")
        Amplify.DataStore.query(Post.self) { result in
            switch result {
            case .success(let posts):
                print("Read result: \(posts)")
                readExpectation2.fulfill()
            case .failure(let error):
                XCTFail("Error reading post \(error)")
            }
        }
        wait(for: [readExpectation2], timeout: 10)
    }
```
The remote DB (DynamoDB) tables are completely empty before running this function. The (trimmed) console output is:
```
Amplify configured with DataStore plugin
...
Clearing local DB
...
Creating post (id=714C3367-199A-4552-8FD4-EA696157951E, title=F4BCD31B-0BF4-4130-B01B-B1502E2031AD)
...
Syncing to cloud
...
Querying posts
Updating post
Deleting post
Syncing to cloud
...
Reading posts
Read result: [DataStoreProtoTests.Post(id: "714C3367-199A-4552-8FD4-EA696157951E", title: "F4BCD31B-0BF4-4130-B01B-B1502E2031AD", status: DataStoreProtoTests.PostStatus.published, rating: Optional(3), content: nil, editors: Optional(Amplify.List<DataStoreProtoTests.PostEditor>), comments: Optional(Amplify.List<DataStoreProtoTests.Comment>))]
```

I would expect the read result to be empty at the end, but as you can see the post which I created is still returned by the query. What I did notice is that the delete does fail on the backend and returns an unresolved conflict error to the client (I have conflict resolution set to "Auto Merge"). If I register a custom `conflictHandler` when configuring DataStore, I do notice that gets called as a result, with both the local and remote versions of the model being identical. If I call `resolve(.retryLocal)` in the conflict handler, this does result in the new record being deleted, which seems to work around this issue. However, I'm worried that this could have undesirable effects in other cases. Please let me know if this behavior is expected, and if this workaround is recommended or not.

**Comment by ruiguoamz** [comment]
Hi, @bgomberg Thanks for reaching out about this.

I found out that by injecting a delay (sleep(1 or 2)) between `update()` and `delete()` that can solve the issue. So I suspect there are some race condition between these two calls. So the workaround (not the expected one) is to inject some delay between them.

At the same time. I need to look deeper inside SDK

**Comment by bgomberg** [comment]
In case it helps, I dumped the `MutationEvent` table right after the delete and noticed that there were two mutation queued, the modify and the delete. I'm not sure if this would completely resolve the issue or if there would still be edge-cases here, but I would have expected the modify mutation to be dropped and just the delete mutation to remain queued.

**Comment by kjones** [comment]
Doesn't seem like pending operations should be aggregated. I would think updates and deletes would all need to go through AppSync due to needing conflict resolution to take place.

**Comment by bgomberg** [comment]
If locally I create and then immediately delete something, as long as nothing is yet synced to the backend, I would expect that the backend would never need to know that happened. Similarly, if I update and then delete something, I don't really care that the updated values make it to the backend as in that case my goal is to have the record be deleted anyways. It looks like there is [code here](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/AWSMutationDatabaseAdapter/AWSMutationDatabaseAdapter+MutationEventIngester.swift#L69) which does this aggregation, but it looks like `localEvents` is specifically ignoring events which have `inProcess` set to true, which makes sense and explains the difference between my "offline" and "online" tests as in the latter case this is set immediately upon saving for the create or update operations. It seems like the solution will involve assuming the operations are not aggregated and better handling the conflict error from the delete operation. Ideally this would happen within AppSync, as it seems pretty clear what the application's intent is in this case.

**Comment by ruiguoamz** [comment]
Hi, @bgomberg 

Thanks for diving deep. 

As you mentioned there are two `MutationEvent` in the `MutationEvent` table, beside the `update` and `delete` mutation type, another important field to notice is the version number which is used for conflict handling. 

I noticed both version number of the update and delete is 1 which should not be case. For update, it should be 1, and for delete, it should be 2. 

I think this the reason of deletion fail, the `MutationEvent` of type update with version 1 is sent to `AppSync`, `AppSync` takes that, applies `Auto Merge` and increments the version to 2. But when `MutationEvent` of type delete is sent, `AppSync` rejects updating of the fields (here means reject marking the `_delete` filed to true) but still incrementing version to 3. 

Reference: [Conflict resolution documentation ](https://docs.aws.amazon.com/appsync/latest/devguide/conflict-detection-and-sync.html)

**Comment by bgomberg** [comment]
@ruiguoamz,

My understanding is that the client shouldn't set the version to 2 until it actually syncs that version down from the backend. The conflict resolution documentation doesn't seem to go into detail on what happens in the delete case for `Auto Merge`, but based on what I'm seeing, it seems like a delete will always result in a `ConflictUnhandled` error if the version doesn't match.

I'm not sure what the best way to fix this within DataStore would be. I suppose DataStore could assume that the version will increase from the update and preemptively set the version for the delete to 2, but that seems error-prone and will fail (potentially rightly-so depending on the desired behavior) if the local DB isn't in sync. It seems like the best solution might be to provide more context to the client's `conflictHandler` to detect these cases. I'm not really sure what to make of the conflict handler being called with the local and remote models being identical from an application perspective.

In general, under what situations is `conflictHandler` called when using the auto merge strategy? I wasn't able to find any documentation describing this. The page you linked only gives examples for the other 2 strategies.

**Comment by palpatim** [comment]
> I noticed both version number of the update and delete is 1 which should not be case. For update, it should be 1, and for delete, it should be 2.

My (1.5 year old) recollection is that we designed locally enqueued mutations to be reconciled at the service. If automerge conflict resolution is set, it *should* apply an update followed by a delete even if the versions in the mutation queue are the same. In no case should we ever be setting the version number locally; those are always controlled by the service.

As @bgomberg mentions, we need to improve documentation on the conflictHandler, and when it's called. We should also look at documenting the reconciliation behavior for various conflict situations.

**Comment by palpatim** [comment]
We are investigating a few different edge cases around consecutive updates, and this is one of them. :) We'll update this issue when we have more info.

**Comment by thisisabhash** [comment]
Changes are merged with #1407. Waiting for release.

**Comment by lawmicha** [comment]
#1407 has been released in https://github.com/aws-amplify/amplify-ios/releases/tag/v1.14.0 This should fix the consecutive update scenarios you've described @bgomberg Will close this for now, please let us know if there are any other issues.
